cmd/cronosd/cmd/fix-unlucky-tx_test.go

```diff
@@ -0,0 +1,131 @@
+package cmd
+
+import (
+	"testing"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/simapp"
+	"github.com/crypto-org-chain/cronos/app"
+	"github.com/stretchr/testify/require"
+	abci "github.com/tendermint/tendermint/abci/types"
+	tmstate "github.com/tendermint/tendermint/proto/tendermint/state"
+	sm "github.com/tendermint/tendermint/state"
+	"github.com/tendermint/tendermint/state/txindex/kv"
+	tmstore "github.com/tendermint/tendermint/store"
+	"github.com/tendermint/tendermint/types"
+	tmdb "github.com/tendermint/tm-db"
+)
+
+func mockResult(txGen client.TxConfig, index uint32, success bool) *abci.TxResult {
+	txs, _ := app.GenSequenceOfTxs(txGen, nil, nil, nil, 1)
+	txBytes, _ := txGen.TxEncoder()(txs[0])
+	result := mockResponseDeliverTx(success)
+	return &abci.TxResult{
+		Height: 1,
+		Index:  index,
+		Tx:     txBytes,
+		Result: *result,
+	}
+}
+
+func mockResponseDeliverTx(success bool) *abci.ResponseDeliverTx {
+	result := abci.ResponseDeliverTx{}
+	if success {
+		result.Code = abci.CodeTypeOK
+		result.Data = []byte{0x01}
+		result.Log = "ok"
+	} else {
+		result.Code = 11
+		result.Log = "out of gas in location: block gas meter; gasWanted:"
+	}
+	return &result
+}
+
+func mockBlockResult() *tmstate.ABCIResponses {
+	return &tmstate.ABCIResponses{
+		DeliverTxs: make([]*abci.ResponseDeliverTx, 0),
+		EndBlock:   &abci.ResponseEndBlock{},
+		BeginBlock: &abci.ResponseBeginBlock{},
+	}
+}
+
+func mockTmDb() *tmDB {
+	db := tmdb.NewMemDB()
+	return &tmDB{
+		blockStore: tmstore.NewBlockStore(db),
+		stateStore: sm.NewStore(db),
+		txIndexer:  kv.NewTxIndex(db),
+	}
+}
+
+type MockTxResult struct {
+	Origin                    *abci.TxResult
+	ReplayedResponseDeliverTx *abci.ResponseDeliverTx
+}
+
+func TestFindUnluckyTx(t *testing.T) {
+	encCfg := simapp.MakeTestEncodingConfig()
+	tmDB := mockTmDb()
+	testCases := []struct {
+		name       string
+		txResults  []MockTxResult
+		expTxIndex int
+	}{
+		{
+			"no unlucky tx",
+			[]MockTxResult{
+				{Origin: mockResult(encCfg.TxConfig, 0, true)},
+				{Origin: mockResult(encCfg.TxConfig, 1, true)},
+				{Origin: mockResult(encCfg.TxConfig, 2, true)},
+			},
+			-1,
+		},
+		{
+			"find unlucky tx",
+			[]MockTxResult{
+				{Origin: mockResult(encCfg.TxConfig, 0, true)},
+				{Origin: mockResult(encCfg.TxConfig, 1, false)},
+				{Origin: mockResult(encCfg.TxConfig, 2, false)},
+			},
+			1,
+		},
+		{
+			"find unlucky tx when indexed as success",
+			[]MockTxResult{
+				{Origin: mockResult(encCfg.TxConfig, 0, true)},
+				{Origin: mockResult(encCfg.TxConfig, 1, false), ReplayedResponseDeliverTx: mockResponseDeliverTx(true)},
```

Comment:
What's the difference with simply `mockResult(encCfg.TxConfig, 1, true)`

Reply:
need write indexer with ok result, what do you mean?

Reply:
I mean it's overridden immediately, it seems no difference from specify an ok result directly?
```
if txResults.ReplayedResponseDeliverTx != nil {
					txResult.Result = *txResults.ReplayedResponseDeliverTx
				}
```

Reply:
put in variable, append blockRes.DeliverTxs 1st,  and [&result](https://github.com/mmsqe/cronos/blob/fix-unlucky-tnx-on-sync-skip/cmd/cronosd/cmd/fix-unlucky-tx_test.go#L117) won

Reply:
So `DeliverTxs` will contain a different result than tx indexer? What's scenario does this try to emulate?

Reply:
txIndexer result is indexed by later block that contains same tnx but success, while current result is still exceedsBlockGasLimit before patch?

Reply:
But when we patch the tx indexer result, we'll patch the block results at the same time, right?

Reply:
I guess you get confused return txIndex is both 2 when change:
`{Origin: mockResult(encCfg.TxConfig, 1, false), ReplayedResponseDeliverTx: mockResponseDeliverTx(true)},`
to
`{Origin: mockResult(encCfg.TxConfig, 1, true)},`
but to test skip case, we need by go into [exceeds](https://github.com/mmsqe/cronos/blob/fix-unlucky-tnx-on-sync-skip/cmd/cronosd/cmd/fix-unlucky-tx.go#L318) condition and modified txIndexer with ok result
update test with require [log](https://github.com/mmsqe/cronos/commit/fd8cdbb4b3b6a73a40e92313f02a52fb194d465e) for these 2 cases